contributors/design-proposals/pod-resource-management.md

```diff
@@ -1,114 +1,419 @@
+allows BestEffort and Burstable pods to potentially consume as many
+resources that are presently available on the node.
+
+For compressible resources, this prioritization scheme has little impact.
+For example, CPU time is proportioned dynamically when there is contention
+using CFS shares that ensure minimum requests are satisfied.
+
+For incompressible resources, this prioritization scheme can inhibit the
+ability of a pod to have its requests satisfied.  For example, a Guaranteed
+pods memory request may not be satisfied if there are active BestEffort
+pods consuming all available memory.
+
+The `kubelet` will support a flag `experimental-qos-reserve-requests` that
+takes a set of percentages per incompressible resource that controls how the
+QoS cgroup sandbox attempts to reserve resources for its tier.  It attempts
+to reserve requested resources to exclude pods from lower OoS classes from
+using resources requested by higher QoS classes. The flag will accept values
+in a range from 0-100%, where a value of 0 instructs the `kubelet` to attempt
+no reservation, and a value of 100 will instruct the `kubelet` to attempt to
+reserve the sum of requested resource across all pods on the node.  How the
+`kubelet` achieves this desired state is resource specific.  
+The default value per incompressible resource if not specified is for no
+reservation to occur for the incompressible resource.
+
+The `kubelet` will allocate resources to the QoS level cgroup
+dynamically in response to the following events:
+
+1. kubelet startup/recovery
+1. prior to creation of the pod level cgroup
+1. after deletion of the pod level cgroup
+1. at periodic intervals to reach `experimental-qos-reserve-requests`
+heurisitcs that converge to a desired state.
+
+All writes to the QoS level cgroup sandboxes are protected via a
+common lock in the kubelet to ensure we do not have multiple concurrent
+writers to this tier in the hierarchy.
+
+### QoS level CPU allocation
+
+The `BestEffort` cgroup sandbox is statically configured as follows:
+
+```
+ROOT/BestEffort/cpu.shares = 2
+```
+
+This ensures that allocation of CPU time to pods in this QoS class
+is given the lowest priority.
+
+The `Burstable` cgroup sandbox CPU share allocation is dynamic based
+on the set of pods currently scheduled to the node.  
+
+```
+ROOT/Burstable/cpu.shares = max(sum(Burstable pods cpu requests, 2)
+```
+
+The Burstable cgroup sandbox is updated dynamically in the exit
+points described in the previous section.  Given the compressible
+nature of CPU, and the fact that cpu.shares are evaluated via relative
+priority, the risk of an update being incorrect is minimized as the `kubelet`
+converges to a desired state.  Failure to set `cpu.shares` at the QoS level
+cgroup would result in `500m` of cpu for a Guaranteed pod to have different
+meaning than `500m` of cpu for a Burstable pod in the current hierarchy.
+For this reason, we will always set `cpu.shares` for the QoS level sandboxes
```

Comment:
I am not following, absent setting a value:

```
$ cat /Burstable/cpu.shares
1024
```

If you have a allocatable=4cpu, and you run the following:

This means if you run stress to do the following:

```
kubectl run burstable --image=sjenning/stress --requests=cpu=500m,memory=100Mi --replicas=1 --command -- /usr/bin/stress -c 4 -t 3600
```

by default, this bursts to consume all 4 cpus, which is fine, that is the point.

but if you do this:

```
kubectl run guaranteed --image=sjenning/stress --requests=cpu=500m,memory=100Mi --limits=cpu=500m,memory=100Mi --replicas=7 --command -- /usr/bin/stress -c 4 -t 3600
```

the result is that the burstable pod skews closer to 1 core of usage since the 1024 share evaluation is made relative to the guaranteed pods, and we really want the Burstable cpu share to be 500 so it was given equal time relative to the other pods.  

i tested this locally and confirmed this was the case, and manually assigning

```
echo 500 > /Burstable/cpu.shares
```

fixes usage to closer to the actual value we requested as expected...

do you disagree that we need to set shares on burstable tier?